source/registry/contracts/Registry.sol

```diff
@@ -16,190 +16,114 @@ contract Registry {
 
   IERC20 public immutable stakingToken;
-  uint256 public immutable obligationCost;
-  uint256 public immutable tokenCost;
+  uint256 public immutable stakingCost;
+  uint256 public immutable supportCost;
   mapping(address => EnumerableSet.AddressSet) internal tokensByStaker;
   mapping(address => EnumerableSet.Bytes32Set) internal protocolsByStaker;
   mapping(address => EnumerableSet.AddressSet) internal stakersByToken;
   mapping(bytes4 => EnumerableSet.AddressSet) internal stakersByProtocol;
   mapping(address => string) public stakerServerURLs;
 
-  event InitialStake(address indexed account);
-  event FullUnstake(address indexed account);
+  event StakeForServer(address indexed account, string url);
+  event AddProtocols(address indexed account, bytes4[] protocols);
   event AddTokens(address indexed account, address[] tokens);
   event RemoveTokens(address indexed account, address[] tokens);
-  event AddProtocols(address indexed account, bytes4[] protocols);
   event RemoveProtocols(address indexed account, bytes4[] protocols);
-  event SetServerURL(address indexed account, string url);
+  event RemoveStakedServer(
+    address indexed account,
+    string url,
+    bytes4[] protocols,
+    address[] tokens
+  );
 
   error NoProtocolsToAdd();
   error NoProtocolsToRemove();
-  error ProtocolDoesNotExist(bytes4);
-  error ProtocolExists(bytes4);
+  error NoServerToRemove();
   error NoTokensToAdd();
   error NoTokensToRemove();
+  error ProtocolDoesNotExist(bytes4);
+  error ProtocolExists(bytes4);
   error TokenDoesNotExist(address);
   error TokenExists(address);
+  error ServerURLInvalid();
 
   /**
    * @notice Constructor
    * @param _stakingToken address of token used for staking
-   * @param _obligationCost base amount required to stake
-   * @param _tokenCost amount required to stake per protocol
+   * @param _stakingCost base amount required to stake
+   * @param _supportCost amount required per token and protocol
    */
   constructor(
     IERC20 _stakingToken,
-    uint256 _obligationCost,
-    uint256 _tokenCost
+    uint256 _stakingCost,
+    uint256 _supportCost
   ) {
     stakingToken = _stakingToken;
-    obligationCost = _obligationCost;
-    tokenCost = _tokenCost;
+    stakingCost = _stakingCost;
+    supportCost = _supportCost;
   }
 
   /**
-   * @notice Set the server URL for a staker
-   * @param _url string value of the ServerURL
+   * @notice Stake tokens and set the server URL
+   * @param _url string value of the URL
    */
-  function setServerURL(string calldata _url) external {
+  function stakeForServer(string calldata _url) external {
+    if (bytes(_url).length == 0) revert ServerURLInvalid();
+
+    if (bytes(stakerServerURLs[msg.sender]).length == 0 && stakingCost > 0) {
+      stakingToken.safeTransferFrom(msg.sender, address(this), stakingCost);
+    }
     stakerServerURLs[msg.sender] = _url;
-    emit SetServerURL(msg.sender, _url);
+    emit StakeForServer(msg.sender, _url);
```

Comment:
Is this a good event name for when we only update the URL?\

Reply:
How about `SetServer(sender, url)` and `UnsetServer(sender, url, protocols, tokens)`?